lib/compiler/passes/inference-match-result.js

```diff
@@ -0,0 +1,110 @@
+"use strict";
+
+const visitor      = require("../visitor");
+const asts         = require("../asts");
+const GrammarError = require("../../grammar-error");
+
+// Inference match result of the each node. Can be:
+// -1: negative result, matching of that node always fails
+//  0: neutral result, may be fail, may be match
+//  1: positive result, always match
+function inferenceMatchResult(ast) {
+  function sometimesMatch(node) { return (node.match = 0); }
+  function alwaysMatch(node) {
+    inference(node.expression);
+
+    return (node.match = 1);
+  }
+
+  function inferenceExpression(node) {
+    return (node.match = inference(node.expression));
+  }
+  function inferenceElements(elements, forChoice) {
+    const length = elements.length;
+    let always = 0;
+    let never = 0;
+
+    for (let i = 0; i < length; ++i) {
+      const result = inference(elements[i]);
+
+      if (result > 0) { ++always; }
+      if (result < 0) { ++never;  }
+    }
+
+    if (always === length) {
+      return 1;
+    }
+    if (forChoice) {
+      return never === length ? -1 : 0;
+    }
+
+    return never > 0 ? -1 : 0;
+  }
+
+  const inference = visitor.build({
+    rule(node) {
+      let oldResult;
+      let count = 0;
+
+      // If property not yet calculated, do that
+      if (typeof node.match === "undefined") {
+        node.match = 0;
+        do {
+          oldResult = node.match;
+          node.match = inference(node.expression);
+          // 6 == 3! -- permutations count for all transitions from one match
+          // state to another.
+          // After 6 iterations the cycle with guarantee begins
+          // istanbul ignore next  This is canary test, shouldn't trigger in real life
+          if (++count > 6) {
+            throw new GrammarError(
+              "Infinity cycle detected when trying to evaluate node match result",
+              node.location
+            );
+          }
+        } while (oldResult !== node.match);
+      }
+
+      return node.match;
+    },
+    named:        inferenceExpression,
```

Comment:
I think it is better to retain the same order as in the `visitor.js` for defaults because then it is easy to visually check that all necessary methods is implemented